sw/applications/spi_flash_write/spi_flash_write.c

```diff
@@ -0,0 +1,274 @@
+    spi_wait_for_ready(&spi_host); 
+
+    // -- DMA CONFIGURATION --
+    dma_set_read_ptr_inc(&dma, (uint32_t) 4); // Do not increment address when reading from the SPI (Pop from FIFO)
+    dma_set_write_ptr_inc(&dma, (uint32_t) 0); // Do not increment address when reading from the SPI (Pop from FIFO)
+    dma_set_read_ptr(&dma, (uint32_t) flash_data); // SPI RX FIFO addr
+    dma_set_write_ptr(&dma, (uint32_t) fifo_ptr); // copy data address
+    // Set the correct SPI-DMA mode:
+    // (0) disable
+    // (1) receive from SPI (use SPI_START_ADDRESS for spi_host pointer)
+    // (2) send to SPI (use SPI_START_ADDRESS for spi_host pointer)
+    // (3) receive from SPI FLASH (use SPI_FLASH_START_ADDRESS for spi_host pointer)
+    // (4) send to SPI FLASH (use SPI_FLASH_START_ADDRESS for spi_host pointer)
+    #ifndef USE_SPI_FLASH
+        dma_set_spi_mode(&dma, (uint32_t) 2); // The DMA will wait for the SPI TX FIFO ready signal
+    #else
+        dma_set_spi_mode(&dma, (uint32_t) 4); // The DMA will wait for the SPI FLASH TX FIFO ready signal
+    #endif
+    dma_set_cnt_start(&dma, (uint32_t) COPY_DATA_BYTES); // Size of data received by SPI
+
+    // Wait for the first data to arrive to the TX FIFO before enabling interrupt
+    spi_wait_for_tx_not_empty(&spi_host);
+    // Enable event interrupt
+    spi_enable_evt_intr(&spi_host, true);
+    // Enable TX empty interrupt
+    spi_enable_txempty_intr(&spi_host, true);
+
+    const uint32_t cmd_write_tx = spi_create_command((spi_command_t){
+        .len        = SPI_BYTES - 1,
+        .csaat      = false,
+        .speed      = kSpiSpeedStandard,
+        .direction  = kSpiDirTxOnly
+    });
+    spi_set_command(&spi_host, cmd_write_tx);
+    spi_wait_for_ready(&spi_host); 
+
+    // Wait for SPI interrupt
+    printf("Waiting for the SPI interrupt...\n");
+    while(spi_intr_flag == 0) {
+        wait_for_interrupt();
+    }
+    printf("triggered!\n");
+
+    // Check status register status waiting for ready
+    bool flash_busy = true;
+    uint8_t flash_resp[4] = {0xff,0xff,0xff,0xff};
+    while(flash_busy){
+        uint32_t flash_cmd = 0x00000005; // [CMD] Read status register
+        spi_write_word(&spi_host, flash_cmd); // Push TX buffer
+        uint32_t spi_bi_cmd = spi_create_command((spi_command_t){
+            .len        = 1,
+            .csaat      = false,
+            .speed      = kSpiSpeedStandard,
+            .direction  = kSpiDirBidir
+        });
+        spi_set_command(&spi_host, spi_bi_cmd);
+        spi_wait_for_ready(&spi_host);
+        spi_wait_for_rx_watermark(&spi_host);
+        spi_read_word(&spi_host, &flash_resp[0]);
+        if ((flash_resp[0] & 0x01) == 0) flash_busy = false;
+    }
+
+    // Fix bug with Bidirectional additional byte
+    spi_set_command(&spi_host, cmd_dummy_read);
+    spi_wait_for_ready(&spi_host);
+    spi_wait_for_rx_watermark(&spi_host);
+    spi_read_word(&spi_host, &flash_resp[0]);
+
+    printf("Power down flash...\n");
```

Comment:
Please check that you actually write to the flash by reading back the values

Reply:
Davide told me not to perform write and read in the same application because it becomes too long, I am checking by reading the entire FLASH with the iceprog yosys program, I will double-check with him

Reply:
@davideschiavone do you agree with this?

Reply:
@benoitdenkinger you're right, I will add the readback to check the values